relay-server/src/services/processor.rs

```diff
@@ -1274,55 +1334,222 @@ impl EnvelopeProcessorService {
         Ok(())
     }
 
-    fn process_state(&self, state: &mut ProcessEnvelopeState) -> Result<(), ProcessingError> {
+    fn process_envelope(
+        &self,
+        managed_envelope: ManagedEnvelope,
+        project_id: ProjectId,
+        project_state: Arc<ProjectState>,
+        sampling_project_state: Option<Arc<ProjectState>>,
+        reservoir_counters: Arc<Mutex<BTreeMap<RuleId, i64>>>,
+    ) -> Result<ProcessingStateResult, (ProcessingError, ManagedEnvelope)> {
         // Get the group from the managed envelope context, and if it's not set, try to guess it
         // from the contents of the envelope.
-        let group = state.managed_envelope.group();
+        let group = managed_envelope.group();
 
         relay_log::trace!("Processing {group:?} group");
 
         match group {
-            ProcessingGroup::Error => self.process_errors(state)?,
-            ProcessingGroup::Transaction => self.process_transactions(state)?,
-            ProcessingGroup::Session => self.process_sessions(state)?,
-            ProcessingGroup::Standalone => self.process_standalone(state)?,
-            ProcessingGroup::ClientReport => self.process_client_reports(state)?,
-            ProcessingGroup::Replay => self.process_replays(state)?,
-            ProcessingGroup::CheckIn => self.process_checkins(state)?,
-            ProcessingGroup::Span => self.process_spans(state)?,
+            ProcessingGroup::Error(ErrorGroup) => {
+                let mut state = self.prepare_state::<ErrorGroup>(
+                    managed_envelope,
+                    project_id,
+                    project_state,
+                    sampling_project_state,
+                    reservoir_counters,
+                );
+                match self.process_errors(&mut state) {
```

Comment:
This is very repetitive, should be able to either do the conversion via `macro_rules` or via `map` and `map_err`.

Skteched up:

```rs
        macro_rules! process {
            ($f:ident) => {{
                let mut state = self.prepare_state(
                    managed_envelope,
                    project_id,
                    project_state,
                    sampling_project_state,
                    reservoir_counters,
                );
                match self.$f(&mut state) {
                    Ok(()) => Ok(ProcessingStateResult {
                        managed_envelope: state.managed_envelope,
                        extracted_metrics: state.extracted_metrics,
                    }),
                    Err(e) => Err((e, state.managed_envelope)),
                }
            }}
        }

        match group {
            ProcessingGroup::Error(_) => process!(process_errors),
            ProcessingGroup::Transaction(_) => process!(process_transactions),
            ...
        }
```

Reply:
Yeah! 
the `map` and `map_err` does not work because of the borrow checker, but `macro_rules` seems good to me